### PR TITLE
[ci] make shell scripts stricter

### PR DESCRIPTION
**Pull request by jameslamb**
Proposes standardizing on 2 practices for shell scripts in this project:

* using `#!/bin/bash` (not `sh` or other shells)
* running shell scripts under stricter settings, to get earlier and louder errors

### Benefits of these changes

Prevents situations where tests are silently failing... like the `rchk` tests have been for over a year (#6332).

Reduces debugging time by ensuring that CI logs tend to end with the first thing that was broken. For example, no more `brew install` silently failing and then only finding out about it via cryptic errors from `R CMD check`.

### Notes for Reviewers

Per the `set` man page ([link](https://linuxcommand.org/lc3_man_pages/seth.html)):

```text
      -e  Exit immediately if a command exits with a non-zero status.
      -o option-name
          Set the variable corresponding to option-name:
              allexport    same as -a
              ...
              pipefail     the return value of a pipeline is the status of
                           the last command to exit with a non-zero status,
                           or zero if no command exited with a non-zero status
              ...
      -u  Treat unset variables as an error when substituting.
```

I've kept the following scripts using `sh`, so they can be executed in the widest possible range of environments (including Windows), without needing to worry about setting up `bash` (which tends to be less common on non-Linux platforms than `sh`):

* `build-python.sh`
* `build-cran-package.sh`
* `.ci/check_python_dists.sh`

### CODE REVIEW

File: .github/workflows/r_package.yml [hunk 1]
```diff
@@ -134,15 +134,6 @@ jobs:
...
             container: null
-          ################
-          # Other checks #
-          ################
-          - os: ubuntu-latest
-            task: r-rchk
-            compiler: gcc
-            r_version: 4.3
-            build_type: cran
-            container: 'ubuntu:22.04'
```
Comment (jameslamb) [inline .github/workflows/r_package.yml:145]: This job has been silently broken for over a year.

I'm working on fixing it in #6332 . See that PR for details.

In this PR, I'm proposing removing it entirely to save some CI time and to not have to introduce complexity to workaround its issues as a part of this PR.

File: .ci/lint-cpp.sh [hunk 2]
```diff
@@ -32,9 +34,12 @@ get_omp_pragmas_without_num_threads() {
         'pragma omp parallel' \
     | grep -v ' num_threads'
 }
+# https://www.gnu.org/software/grep/manual/html_node/Exit-Status.html
+set +e +o pipefail
```
Comment (borchero) [inline .ci/lint-cpp.sh:38]: Could you add a brief description in prose why this is necessary? It took me a second to see that `get_omp_pragmas_without_num_threads` runs `grep`

File: build-cran-package.sh
Comment (borchero) [inline build-cran-package.sh:1]: This file uses `#!/bin/sh` still

File: .ci/check_python_dists.sh [hunk 3]
```diff
@@ -1,7 +1,13 @@
 #!/bin/sh
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply .ci/check_python_dists.sh:1]: No. From the description of the PR:

> *I've kept the following scripts using `sh`, so they can be executed in the widest possible range of environments (including Windows), without needing to worry about setting up `bash` (which tends to be less common on non-Linux platforms than sh):*
>
> - `build-python.sh`
> - `build-cran-package.sh`
> - `.ci/check_python_dists.sh`

File: build-cran-package.sh
Comment (jameslamb) [reply build-cran-package.sh:1]: This is intentional, see https://github.com/microsoft/LightGBM/pull/6266#discussion_r1575362174

File: .ci/check_python_dists.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (borchero) [reply .ci/check_python_dists.sh:1]: Ahh sorry, I only read the first two paragraphs 😅

[the remark this answers is not in the document]
Comment (jameslamb) [reply .ci/check_python_dists.sh:1]: No problem, thanks for your thorough review!

File: .ci/lint-cpp.sh (hunk 2 above)
Comment (jameslamb) [reply .ci/lint-cpp.sh:43]: updated in https://github.com/microsoft/LightGBM/pull/6266/commits/e4fa87d2743ee91ccf8b89c43a54e212f6971183, let me know what you think

Comment (borchero) [reply .ci/lint-cpp.sh:43]: Reads nicely, I think it should be "returns a non-0 exit code if **no** lines were found" though, right? 🤔

Comment (jameslamb) [reply .ci/lint-cpp.sh:43]: ah!! yes you're totally right, sorry 😓

File: .ci/lint-cpp.sh [hunk 4]
```diff
@@ -32,9 +34,17 @@ get_omp_pragmas_without_num_threads() {
         'pragma omp parallel' \
     | grep -v ' num_threads'
 }
+
+# 'grep' returns a non-0 exit code if any lines were found.
```
Comment (jameslamb) [inline .ci/lint-cpp.sh:38]:
```suggestion
# 'grep' returns a non-0 exit code if 0 lines were found.
```

File: .ci/lint-cpp.sh (hunk 2 above)
Comment (jameslamb) [reply .ci/lint-cpp.sh:43]: fixed in https://github.com/microsoft/LightGBM/pull/6266/commits/9d24415f0423fb4ab4e245bfa0a6575d8af3ad82

**Review by borchero — APPROVED** [review]
Nice, thanks!

**Pull request merged**